core/string/translation.cpp

```diff
@@ -1300,6 +1333,275 @@ StringName TranslationServer::doc_translate_plural(const StringName &p_message,
+		if (pseudolocalization_skip_placeholders_enabled && message[i] == '%' && i < message.size() - 1 &&
+				(message[i + 1] == 's' || message[i + 1] == 'c' || message[i + 1] == 'd' ||
+						message[i + 1] == 'o' || message[i + 1] == 'x' || message[i + 1] == 'X' || message[i + 1] == 'f')) {
+			res += message[i];
+			res += message[i + 1];
+			i++;
+			continue;
+		}
+		res += message[i];
+		if (message[i] == 'a' || message[i] == 'e' || message[i] == 'i' || message[i] == 'o' || message[i] == 'u' ||
+				message[i] == 'A' || message[i] == 'E' || message[i] == 'I' || message[i] == 'O' || message[i] == 'U') {
+			res += message[i];
+		}
+	}
+	return res;
+};
+
+String TranslationServer::replace_with_accented_string(String &message) const {
+	String res = "";
+	for (int i = 0; i < message.size(); i++) {
+		if (pseudolocalization_skip_placeholders_enabled && message[i] == '%' && i < message.size() - 1 &&
+				(message[i + 1] == 's' || message[i + 1] == 'c' || message[i + 1] == 'd' ||
+						message[i + 1] == 'o' || message[i + 1] == 'x' || message[i + 1] == 'X' || message[i + 1] == 'f')) {
+			res += message[i];
+			res += message[i + 1];
+			i++;
+			continue;
+		}
+		const char32_t *accented = get_accented_version(message[i]);
+		if (accented) {
+			res += accented;
+		} else {
+			res += message[i];
+		}
+	}
+	return res;
+}
+
+String TranslationServer::wrap_with_fakebidi_characters(String &message) const {
+	String res = "";
+	char32_t fakebidiprefix = U'\u202e';
+	char32_t fakebidisuffix = U'\u202c';
+	res += fakebidiprefix;
+	//the fake bidi unicode gets popped at every newline so pushing it back at every newline.
+	for (int i = 0; i < message.size(); i++) {
+		if (message[i] == '\n') {
+			res += fakebidisuffix;
+			res += message[i];
+			res += fakebidiprefix;
+		} else if (pseudolocalization_skip_placeholders_enabled && message[i] == '%' && i < message.size() - 1 &&
+				   (message[i + 1] == 's' || message[i + 1] == 'c' || message[i + 1] == 'd' ||
+						   message[i + 1] == 'o' || message[i + 1] == 'x' || message[i + 1] == 'X' || message[i + 1] == 'f')) {
+			res += fakebidisuffix;
+			res += message[i];
+			res += message[i + 1];
+			res += fakebidiprefix;
+			i++;
+		} else {
+			res += message[i];
+		}
+	}
+	res += fakebidisuffix;
+	return res;
+}
+
+String TranslationServer::add_padding(String &message, int length) const {
+	String res = "";
+	String prefix = pseudolocalization_prefix;
+	String suffix = "";
+	for (int i = 0; i < length * expansion_ratio / 2; i++) {
+		prefix += "_";
+		suffix += "_";
+	}
+	suffix += pseudolocalization_suffix;
+	//replace with expansion logic here when implementing that
+	res += prefix;
+	res += message;
+	res += suffix;
+	return res;
+}
+
+const char32_t *TranslationServer::get_accented_version(char32_t c) const {
+	switch (c) {
+		case 'A':
+			return U"Å";
+		case 'B':
+			return U"ß";
+		case 'C':
+			return U"Ç";
+		case 'D':
+			return U"Ð";
+		case 'E':
+			return U"É";
```

Comment:
I wondered if this could be made a bit more compact with some kind of lookup table instead of manual switch.